lib/completion.ts

```diff
@@ -351,9 +356,32 @@ export class Completion implements CompletionInstance {
       : templates.completionShTemplate;
     const name = this.shim.path.basename($0);
 
+    // Sanitize `name` to make sure it is a valid bash function name by
+    // replacing any characters that are not alphanumeric or underscore with an
+    // underscore
+    const safeName = name.replace(/[^A-Za-z0-9_]/g, '_');
+
     // add ./ to applications not yet installed as bin.
     if ($0.match(/\.js$/)) $0 = `./${$0}`;
 
+    // zsh supports multi-word completions by default, but bash needs custom logic
+    const appNameWords = name.split(' ');
+    if (!this.zshShell) {
+      // to support multi-word completion check each word in the app name
+      const appNameFirstWord = name[0];
+      const wordsToCheck = appNameWords
+        .map((word, index) => {
+          return `"\${COMP_WORDS[${index}]}" == "${word}"`;
+        })
+        .join(' && ');
+      script = script.replace(/{{app_name_first_word}}/g, appNameFirstWord);
+      script = script.replace(/{{app_name_word_check}}/g, wordsToCheck);
+    } else {
+      // zsh multi-word completion works by replacing the space with a '='
+      script = script.replace(/{{zsh_app_name}}/g, appNameWords.join('='));
+    }
+
+    script = script.replace(/{{safe_app_name}}/g, safeName);
```

Comment:
To save you from having to figure it out. For a script name of "yarn webpack" it'd create this:

```bash
_yarn_webpack_yargs_completions()
{
    if [[ "${COMP_WORDS[0]}" == "yarn" && "${COMP_WORDS[1]}" == "webpack" ]]; then
        local cur_word args type_list

        cur_word="${COMP_WORDS[COMP_CWORD]}"
        args=("${COMP_WORDS[@]}")

        # ask yargs to generate completions.
        type_list=$(yarn webpack --get-yargs-completions "${args[@]}")

        COMPREPLY=( $(compgen -W "${type_list}" -- ${cur_word}) )

        # if no match was found, fall back to filename completion
        if [ ${#COMPREPLY[@]} -eq 0 ]; then
          COMPREPLY=()
        fi
    fi

    return 0
}
complete -o bashdefault -o default -F _yarn_webpack_yargs_completions yarn
```

and for zsh it'd create:

```bash
_yarn_webpack_yargs_completions()
{
  local reply
  local si=$IFS
  IFS=$'
' reply=($(COMP_CWORD="$((CURRENT-1))" COMP_LINE="$BUFFER" COMP_POINT="$CURSOR" yarn webpack --get-yargs-completions "${words[@]}"))
  IFS=$si
  _describe 'values' reply
}
compdef _yarn_webpack_yargs_completions "yarn=webpack"
```